File: practical-09-rail-fence-cipher/rail_fence_cipher.py
def rail_fence_encrypt(plaintext, rails):
    """Encrypt plaintext using Rail Fence transposition cipher."""
    if rails <= 1 or len(plaintext) <= rails:
        return plaintext

    fence = [[] for _ in range(rails)]
    rail = 0
    direction = 1

    for char in plaintext:
        fence[rail].append(char)

        if rail == 0:
            direction = 1
        elif rail == rails - 1:
            direction = -1

        rail += direction

    result = ""
    for row in fence:
        result += "".join(row)

    return result


def rail_fence_decrypt(ciphertext, rails):
    """Decrypt ciphertext from Rail Fence transposition cipher."""
    if rails <= 1 or len(ciphertext) <= rails:
        return ciphertext

    n = len(ciphertext)
    fence = [[0] * n for _ in range(rails)]

    rail = 0
    direction = 1
    for i in range(n):
        fence[rail][i] = 1

        if rail == 0:
            direction = 1
        elif rail == rails - 1:
            direction = -1

        rail += direction

    chars_index = 0
    for r in range(rails):
        for c in range(n):
            if fence[r][c] == 1:
                fence[r][c] = ciphertext[chars_index]
                chars_index += 1

    result = ""
    rail = 0
    direction = 1
    for i in range(n):
        result += fence[rail][i]

        if rail == 0:
            direction = 1
        elif rail == rails - 1:
            direction = -1

        rail += direction

    return result


if __name__ == "__main__":
    text = "HELLO WORLD"
    key = 3

    encrypted = rail_fence_encrypt(text, key)
    decrypted = rail_fence_decrypt(encrypted, key)

    print(f"Original: {text}")
    print(f"Rails: {key}")
    print(f"Encrypted: {encrypted}")
    print(f"Decrypted: {decrypted}")

File: c programs/rail_fence_cipher.c
#include <stdio.h>
#include <string.h>

void rail_fence_encrypt(const char *plaintext, int rails, char *result) {
    if (rails <= 1 || strlen(plaintext) <= rails) {
        strcpy(result, plaintext);
        return;
    }
    
    int len = strlen(plaintext);
    char fence[100][100];
    int fence_len[100] = {0};
    
    for (int i = 0; i < rails; i++) {
        for (int j = 0; j < len; j++) {
            fence[i][j] = '\0';
        }
    }
    
    int rail = 0;
    int direction = 1;
    
    for (int i = 0; i < len; i++) {
        fence[rail][fence_len[rail]] = plaintext[i];
        fence_len[rail]++;
        
        if (rail == 0) {
            direction = 1;
        } else if (rail == rails - 1) {
            direction = -1;
        }
        rail += direction;
    }
    
    int r_idx = 0;
    for (int i = 0; i < rails; i++) {
        for (int j = 0; j < fence_len[i]; j++) {
            result[r_idx] = fence[i][j];
            r_idx++;
        }
    }
    result[r_idx] = '\0';
}

void rail_fence_decrypt(const char *ciphertext, int rails, char *result) {
    if (rails <= 1 || strlen(ciphertext) <= rails) {
        strcpy(result, ciphertext);
        return;
    }
    
    int len = strlen(ciphertext);
    int fence[100][100];
    
    for (int i = 0; i < rails; i++) {
        for (int j = 0; j < len; j++) {
            fence[i][j] = 0;
        }
    }
    
    int rail = 0;
    int direction = 1;
    for (int i = 0; i < len; i++) {
        fence[rail][i] = 1;
        
        if (rail == 0) {
            direction = 1;
        } else if (rail == rails - 1) {
            direction = -1;
        }
        rail += direction;
    }
    
    int chars_index = 0;
    for (int r = 0; r < rails; r++) {
        for (int c = 0; c < len; c++) {
            if (fence[r][c] == 1) {
                fence[r][c] = ciphertext[chars_index];
                chars_index++;
            }
        }
    }
    
    int r_idx = 0;
    rail = 0;
    direction = 1;
    for (int i = 0; i < len; i++) {
        result[r_idx] = (char)fence[rail][i];
        r_idx++;
        
        if (rail == 0) {
            direction = 1;
        } else if (rail == rails - 1) {
            direction = -1;
        }
        rail += direction;
    }
    result[r_idx] = '\0';
}

int main() {
    const char *text = "HELLO WORLD";
    int key = 3;
    
    char encrypted[100];
    char decrypted[100];
    
    rail_fence_encrypt(text, key, encrypted);
    rail_fence_decrypt(encrypted, key, decrypted);
    
    printf("Original: %s\n", text);
    printf("Rails: %d\n", key);
    printf("Encrypted: %s\n", encrypted);
    printf("Decrypted: %s\n", decrypted);
    
    return 0;
}
